R/do_findChromPeaks-functions.R

```diff
@@ -2220,9 +2220,220 @@ do_findPeaks_MSW <- function(mz, int, snthresh = 3,
+                   verboseColumns = verboseColumns, roiList = roiList,
+                   firstBaselineCheck = firstBaselineCheck,
+                   roiScales = roiScales, sleep = sleep,
+                   extendLengthMSW = extendLengthMSW)
+    
+    ## Resolving peaks across IM dimension
+    message("Resolving peaks over ion-mobility dimension... ", appendLF = F)
+    resolved_peaks <- vector("list", nrow(peaks))
+    for (i in seq_len(nrow(peaks))) {
+        current_peak <- peaks[i,]
+        mobilogram <- .extract_mobilogram(pdata, current_peak, rt, im, binWidthIM)
+        if (length(mobilogram) == 0) {
+            warning(i, " mobilogram is empty")
+            next
+        }
+        bounds <- .split_mobilogram(mobilogram)
+        new_peaks <- data.frame(
+            mz = current_peak["mz"],
+            mzmin = current_peak["mzmin"],
+            mzmax = current_peak["mzmax"],
+            rt = current_peak["rt"],
+            rtmin = current_peak["rtmin"],
+            rtmax = current_peak["rtmax"],
+            im = vapply(bounds, mean, numeric(1)),
+            immin = vapply(bounds, min, numeric(1)),
+            immax = vapply(bounds, max, numeric(1)),
+            row.names = NULL
+        )
+        resolved_peaks[[i]] <- new_peaks
+    }
+    resolved_peaks <- do.call(rbind, resolved_peaks)
+    message("OK")
+    
+    ## Refine and calculate peak parameters
+    vals <- vector("list", nrow(resolved_peaks))
+    for (i in seq(nrow(resolved_peaks))) {
+        peak <- unlist(resolved_peaks[i, , drop = TRUE])
+        
+        ## Create a EIC for mz, rt and IM ranges
+        eic <- .extract_EIC_IM(peak, pdata, rt, im)
+        
+        if (nrow(eic) == 0 | all(eic[, 2] == 0))
+            next
+        
+        ## Refine RT bounds
+        rts <- c(peak["rtmin"], peak["rtmax"])
+        apx <- which.max(eic[, 2])
+        apx_rt <- eic[apx, 1]
+        range <- xcms:::descendMin(eic[, 2], apx)
+        
+        eic <- eic[range[1]:range[2], , drop = FALSE]
+        
+        ## Calculate peak stats
+        vals[[i]] <- data.frame(
+            mz = peak["mz"],
+            mzmin = peak["mzmin"],
+            mzmax = peak["mzmax"],
+            rt = apx_rt,
+            rtmin = min(eic[, 1]),
+            rtmax = max(eic[, 1]),
+            im = peak["im"],
+            immin = peak["immin"],
+            immax = peak["immax"],
+            maxo = max(eic[, 2]),
+            into = sum(eic[, 2]),
+            intb = NA,
+            sn = NA
+        )
+    }
+    resolved_peaks <- do.call(rbind, vals)
+    resolved_peaks <-
+        resolved_peaks[resolved_peaks$into > 0, ] #Remove empty peaks
+    
+    as.matrix(resolved_peaks)
+}
+
+#' @importFrom MsCoreUtils bin
+.extract_mobilogram <- function(pdata, peak, rt, im, binWidthIM = 0.01){
+    rtr <- c(peak["rtmin"], peak["rtmax"])
+    mzr <- c(peak["mzmin"], peak["mzmax"])
+    keep <- dplyr::between(rt, rtr[1], rtr[2])
```

Comment:
Why not using `MsCoreUtils::between` instead?

Reply:
I also tried it, but the performance was much better for `dplyr::between`, around 2 times faster
Still, since rt is a sorted vector (by default I think, Spectra wouldn't allow it otherwise), using `base::findInterval` would be even faster. Here's a representative benchmark:

```
library(microbenchmark)
rt <- rep(1:2000, each = 1000)  ## Big vector, like the ones you'd have for IM scans
rtint <- c(230, 260)
microbenchmark(MsCoreUtils::between(rt, rtint),
                            dplyr::between(rt, rtint[1], rtint[2]),
                            findInterval(rtint, rt), times = 1000)
```
![image](https://user-images.githubusercontent.com/45530914/206308281-d1c45fb5-5d6c-4439-9712-82edfe111d75.png)

Reply:
wow - indeed. note that `rtime` does not necessarily need to be ordered - `Spectra` is relatively relaxed there (while `mz` values indeed are required to be sorted). I'll open an issue in `MsCoreUtils` - no need to have a `between` function there if there is a more powerful in `dplyr`...

Reply:
I see, I got it the other way around with the `rtime` and `mz`, I thought `rtime` also had to be ordered: then, could we just order the spectra by `rtime` first so we can use the faster function `findInterval`? 
Something like this, at some point: 
`spec <- spec[order(rtime(spec))]`

I like the fact that the `MsCoreUtils::between` uses a `numeric(2)` as input and sorts it before using it (it's neat and avoids silly mistakes). I'll comment that on the corresponding issue

Reply:
Actually, I think it would be best to start using the `MsCoreUtils::between` again - Sebastian did some improvement in a C implementation of the code. So it should now be comparable (and eventually even faster) than `dplyr::between`.

Regarding `findInterval` and ordering the `rtime` - you're right, rts *should* be ordered, but I would like to avoid checking for the order or enforcing an order... better to have the code as independent of possible from potential pre-requisites (that we would then always need to check if they are true within the function).

Reply:
Seems good! I'll revert back to `MsCoreUtils::between` then :+1: